cmd/kubeadm/app/discovery/file/file.go

```diff
@@ -30,37 +34,69 @@ import (
+func RetrieveValidatedConfigInfo(filepath, clustername string) (*clientcmdapi.Config, error) {
+	config, err := clientcmd.LoadFromFile(filepath)
 	if err != nil {
 		return nil, err
 	}
-	return ValidateClusterInfo(clusterinfo, clustername)
+	return ValidateConfigInfo(config, clustername)
 }
 
-// ValidateClusterInfo connects to the API Server and makes sure it can talk
-// securely to the API Server using the provided CA cert and
+// ValidateConfigInfo connects to the API Server and makes sure it can talk
+// securely to the API Server using the provided CA cert/client certificates  and
 // optionally refreshes the cluster-info information from the cluster-info ConfigMap
-func ValidateClusterInfo(clusterinfo *clientcmdapi.Config, clustername string) (*clientcmdapi.Cluster, error) {
-	err := validateClusterInfoKubeConfig(clusterinfo)
+func ValidateConfigInfo(config *clientcmdapi.Config, clustername string) (*clientcmdapi.Config, error) {
+	err := validateKubeConfig(config)
 	if err != nil {
 		return nil, err
 	}
 
 	// This is the cluster object we've got from the cluster-info KubeConfig file
-	defaultCluster := kubeconfigutil.GetClusterFromKubeConfig(clusterinfo)
+	defaultCluster := kubeconfigutil.GetClusterFromKubeConfig(config)
+
+	var authInfo *clientcmdapi.AuthInfo
+	if config.Contexts[config.CurrentContext] != nil {
+		// load pre-existing client certificates
+		authInfo = config.AuthInfos[config.Contexts[config.CurrentContext].AuthInfo]
+		if len(authInfo.ClientCertificateData) == 0 && len(authInfo.ClientCertificate) != 0 {
+			clientCert, err := extractFromFile(authInfo.ClientCertificate)
+			if err != nil {
+				return nil, err
+			}
+			authInfo.ClientCertificateData = clientCert
+		}
+		if len(authInfo.ClientKeyData) == 0 && len(authInfo.ClientKey) != 0 {
+			clientKey, err := extractFromFile(authInfo.ClientKey)
+			if err != nil {
+				return nil, err
+			}
+			authInfo.ClientKeyData = clientKey
+		}
+	}
 
 	// Create a new kubeconfig object from the given, just copy over the server and the CA cert
 	// We do this in order to not pick up other possible misconfigurations in the clusterinfo file
-	configFromClusterInfo := kubeconfigutil.CreateBasic(
-		defaultCluster.Server,
-		clustername,
-		"", // no user provided
-		defaultCluster.CertificateAuthorityData,
-	)
+	var configFromClusterInfo *clientcmdapi.Config
+	if len(authInfo.ClientCertificateData) > 0 && len(authInfo.ClientKeyData) > 0 {
+		configFromClusterInfo = kubeconfigutil.CreateWithCerts(
+			defaultCluster.Server,
+			clustername,
+			"", // no user provided
+			defaultCluster.CertificateAuthorityData,
+			authInfo.ClientKeyData,
+			authInfo.ClientCertificateData,
+		)
+	} else {
+		configFromClusterInfo = kubeconfigutil.CreateBasic(
+			defaultCluster.Server,
+			clustername,
+			"", // no user provided
+			defaultCluster.CertificateAuthorityData,
+		)
+	}
 
 	client, err := kubeconfigutil.ToClientSet(configFromClusterInfo)
```

Comment:
You should/can now remove this code that looks up the cluster info ConfigMap, not needed anymore. It could make sense to do that in a separate PR that would be merged first, but I don't have a strong preference here.